cuqi/distribution/_gaussian.py

```diff
@@ -1,31 +1,43 @@
+import warnings
+import numbers
 import numpy as np
+import numpy.linalg as nplinalg
+
+import scipy as sp
 import scipy.stats as sps
-from scipy.sparse import diags, identity, issparse, vstack
-from scipy.sparse import linalg as splinalg
-from scipy.linalg import eigh, eigvals, cholesky
+import scipy.sparse as spa
+import scipy.linalg as splinalg
+
+# from scipy.sparse import diags, identity, issparse, vstack
+# from scipy.sparse import linalg as splinalg
+# from scipy.linalg import eigh, eigvals, cholesky
+
+try:
+    import sksparse.cholmod as skchol
+    chols = True
+except ImportError:
+    chols = False
+
+from cuqi import config
 from cuqi.geometry import _get_identity_geometries
 from cuqi.utilities import force_ndarray, sparse_cholesky
-from cuqi import config
-import warnings
-import numbers
 from cuqi.distribution import Distribution
 
-class GaussianCov(Distribution): # TODO: super general with precisions
+# TODOs:
+# Support full sparse covariance matrices without cholmod library
+
+class GaussianCov(Distribution):
     """
-    General Gaussian probability distribution. Generates instance of cuqi.distribution.GaussianCov
+    General Gaussian probability distribution. Generates instance of cuqi.distribution.Gaussian
 
-    
     Parameters
     ------------
     mean: Mean of distribution. Can be a scalar or 1d numpy array
-    cov: Covariance of distribution. Can be a scalar, 1d numpy array (assumes diagonal elements), or 2d numpy array.
     
-    Methods
-    -----------
-    sample: generate one or more random samples
-    pdf: evaluate probability density function
-    logpdf: evaluate log probability density function
-    cdf: evaluate cumulative probability function
+    cov: Covariance of the distribution. Can be a scalar, 1d numpy array (assumes diagonal elements), or 2d numpy array.
+    prec: Precision of the distribution. Can be a scalar, 1d numpy array (assumes diagonal elements), or 2d numpy array.
+    sqrtcov: A matrix R, where R.T@R = CovarianceMatrix of the distribution. Can be a 2d sparse or numpy array.
+    sqrtprec: A matrix R, where R.T@R = PrecisionMatrix of the distribution. Can be a 2d sparse or numpy array.  
```

Comment:
Can the prec and cov not be 2d sparse array?

Reply:
They can if cholmod is installed, otherwise we get an error. I will re-write this a bit